### PR TITLE
feat: add field-level permissions support

### PR DESCRIPTION
**Pull request by sarasla**
Tests should be added to cover this change.

### CODE REVIEW

**Comment by philjoseph** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [4fbe3018\-1b82\-4033\-b4d2\-0557ca06ad2a](https://deu.ast.checkmarx.net/projects/d7d6dd65-9a7a-4a04-8b7c-54d3aadf247b/scans?id=4fbe3018-1b82-4033-b4d2-0557ca06ad2a&branch=fixing-error-of-playwright)
### Great job! No new security vulnerabilities introduced in this pull request


---
Communicate with Checkmarx by submitting a PR comment with @Checkmarx followed by one of the supported commands. Learn about the supported commands [here](https://docs.checkmarx.com/en/34965-68683-code-repository-integration-usage---results.html#UUID-40215a0c-63e4-58a7-7a20-eb8732bc3d5e_section-idm235086038509635).

File: packages/authorization/src/access/general.ts [hunk 1]
```diff
@@ -30,17 +32,33 @@ export const canUserAccessAction = async (
...
+              hasFieldAccess = true;
+            }
+          }
         }
       }
     }
   }
-  return userAllowedActions.has(action);
+
+  return userAllowedActions.has(action) && hasFieldAccess;
```
Comment (tsemachh) [inline packages/authorization/src/access/general.ts]: **Potential logic inconsistency across roles:** `hasFieldAccess` and `userAllowedActions` are accumulated independently. A scenario where Role A grants entity access (populating `userAllowedActions`) but has field restrictions, and Role B grants the action without matching the entity, could leave `hasFieldAccess = false` while `userAllowedActions.has(action) = true`. Consider evaluating both conditions together per-role, or at minimum add a comment explaining the intended multi-role behavior.

File: packages/authorization/src/types.ts [hunk 2]
```diff
@@ -1,5 +1,11 @@
 import { CollectionSlug } from 'payload';
```
Comment (tsemachh) [inline packages/authorization/src/types.ts:2]: `FieldLevelPermission` is exported but never used to type the `permission` variable in `general.ts`. Consider using it: `const permission: FieldLevelPermission` inside the loop, or remove it if it's only aspirational.

File: packages/authorization/src/access/general.ts [hunk 3]
```diff
@@ -7,12 +7,14 @@ const PERMISSION_HIERARCHY: Record<string, string[]> = {
...
+// Returns true if user has action access to the collection, and (if fieldName provided) to the field
 export const canUserAccessAction = async (
   user: User | null | undefined,
   slugName: string,
   action: string,
   payload: Payload,
   config: AuthorizationPluginConfig,
+  fieldName?: string,
 ): Promise<boolean> => {
   if (!user) return false;
```
Comment (tsemachh) [inline packages/authorization/src/access/general.ts:21]: The new `fieldName` parameter changes the public API of `canUserAccessAction`. Please check all existing call sites to ensure they are not broken, and consider adding a JSDoc comment documenting the parameter's behavior (optional, defaults to allowing all fields).

**Review by tsemachh — CHANGES_REQUESTED** [review]
## Code Review

### Summary
This PR introduces **field-level permission support** to the authorization package and fixes a bug where `permission.type` was incorrectly treated as a `string` instead of `string[]` (a `hasMany` select field). It also fixes missing `.js` extensions in ESM imports. These are meaningful improvements, but there are several issues to address before merging.

---

### 🔴 Critical Issues

**1. PR title is misleading**
The title `fixing-error-of-playwright` does not reflect the actual changes. This is a feature addition (field-level permissions) + a bug fix (`permission.type` as `string[]`). Please rename the PR title to something accurate like `feat: add field-level permissions support`.

**2. No tests**
The PR description itself acknowledges: _"Tests should be added to cover this change."_ The new `fieldName` parameter and the field-level access logic in `canUserAccessAction` are untested. This is a security-sensitive access control function — it must have test coverage before merging. Please add unit tests covering at minimum:
- `fieldName` not provided → behaves as before (backward compatibility)
- `fieldName` provided, field is in `permission.fields` → returns `true`
- `fieldName` provided, field is NOT in `permission.fields` → returns `false`
- `permission.fields` is empty → all fields allowed
- Multiple permission types (hasMany) with hierarchy

**3. `hasFieldAccess` logic has a subtle bug**
```ts
let hasFieldAccess = !fieldName; // If no fieldName provided, default to true
```
Then inside the loop:
```ts
if (fieldName && types.some((type) => PERMISSION_HIERARCHY[type]?.includes(action))) {
```
This means `hasFieldAccess` is only set to `true` if the entity slug matches AND the type hierarchy includes the action. However, if a user has multiple roles and one role grants entity access without field restriction, `hasFieldAccess` will correctly become `true`. But if the entity match happens in a role that does NOT include the action in its hierarchy, `hasFieldAccess` stays `false` even though `userAllowedActions` may have been populated by another role. The two checks (`userAllowedActions` and `hasFieldAccess`) are accumulated independently across roles, which can lead to inconsistent results. Consider tracking field access per-role or restructuring the logic so both checks are evaluated together.

---

### 🟡 Minor Issues

**4. `FieldLevelPermission` type in `types.ts` is defined but not used**
The new `FieldLevelPermission` type is exported but never imported or used in `general.ts` or `Roles.ts`. Either use it to type `permission` in the loop, or remove it to avoid dead code.

**5. Width change in `Roles.ts` is undocumented**
The `entity` field width was changed from `70%` to `40%` to accommodate the new `fields` column. This is fine, but the three columns (`entity` at 40%, `type` at unknown%, `fields` at 30%) should add up sensibly. Please verify the layout is intentional and document it if needed.

**6. Trailing blank lines removed**
Removing trailing newlines at the end of `Roles.ts` is a cosmetic change — fine to include, but worth noting it's unrelated to the feature.

---

### ✅ What's Good
- Fixing `permission.type` from `string` to `string[]` is a correct and necessary bug fix.
- Adding `.js` extensions to ESM imports (`isAdmin.js`, `isAdminOrSelf.js`, `isLoggedIn.js`) is the right fix for ESM module resolution.
- The `fields` field in the `Roles` collection is well-structured (`hasMany: true`, localized label, clear description).
- The comment `// No fields specified = all fields allowed` clearly documents the intended behavior.

---

### Action Items
- [ ] Rename the PR title
- [ ] Add unit/integration tests for `canUserAccessAction` with `fieldName`
- [ ] Fix or clarify the cross-role `hasFieldAccess` accumulation logic
- [ ] Use or remove the `FieldLevelPermission` type
